### ISSUE TITLE
getVisibility for KSDeclaration returns java package private as public in .class dependencies

### ISSUE THREAD

**Issue by yigit** [google/ksp#159]
When a code like the following is inherited via classpath (comes from a dependency), the field (`javaPackage`) is reported as public instead of javaPackagePrivate from `KSDeclaration.getVisibility`.

```
public class JavaClass {
    Long javaPackage;
}
```
It has the side effect of making `KSDeclaration.isPublic` return `true` instead of `false` as well.
It is probably related to [this](https://github.com/google/ksp/blob/master/api/src/main/kotlin/com/google/devtools/ksp/utils.kt#L119) line.

**Comment by yigit** [comment]
Seems like there is also a problem for kotlin properties.

For the following input:
```
// FILE: KotlinC.kt
class KotlinC {
    val x:Int = 0
}

// FILE: JavaC.java
class JavaC {
    private int x;
    public void setX(int x) {}
    public int getX() { return x; }
}
```

I would expect the `x` property to have the same modifiers returned but they are different:
```
KotlinC: public
KotlinC.x: public

JavaC: open, javaPackagePrivate
JavaC.x: open, private
```

This will be challenging when generating kotlin code as normally one can call `JavaC.x` as if it is a property but KSP reports it as private. Even more, looks like `KSPropertyDeclarationJavaImpl `does not implement `getter/setter` so it won't look like a proper property from KSP code. (is probably a separate but related bug). That means, the symbol processor must implement their own getter setter logic for it.

I'll keep digging, just wanted to share some findings. It is possible that i have really unreasonable expectations here :shrug:

**Comment by neetopia** [comment]
Actually I think it is possible to check if the accessor in Java belongs to a property in same class, but the problem here is that we might be doing too much for this. For example, we will need to lift the aforementioned accessors out of `JavaC.declarations`, since accessors belongs to the property, not the class `JavaC`

**Comment by yigit** [comment]
I'm surprised that the compiler does not do that for us already.

Thinking more about this, if we don't do that work, it would require every developer to account for getters/setters (and find them wrt kotlin's rules) if the origin is java.

**Comment by yigit** [comment]
also for reference, kotlin-java interop docs lists these as properties:

https://kotlinlang.org/docs/reference/java-interop.html#getters-and-setters

```
Methods that follow the Java conventions for getters and setters (no-argument methods with names starting with get and single-argument methods with names starting with set) are represented as properties in Kotlin.
```

**Comment by ZacSweers** [comment]
auto-common's `Visibility.java` class might have some good inspiration for this https://github.com/google/auto/blob/master/common/src/main/java/com/google/auto/common/Visibility.java#L74. I think it would be good to indicate package-private for java sources only but push people toward a `KSDeclaration.isVisibleTo(other: KSElement)` or similar API

**Comment by neetopia** [comment]
fixed in #600 , note that for java property accessors, it is more tricky as there is no guarantee that the function with the naming convention is associated with a java field, therefore we decide to not handle this case for now(user can still write custom logic to handle this if needed)
